### ISSUE TITLE
Chaining where statements produces incorrect SQL

### ISSUE THREAD

**Issue by spiffistan** [rails/rails#23055]
We recently upgraded to Rails 4.2.5 (from latest rails 3.2.x) and noticed something strange in some of our statements.

The following code will produce incorrect SQL with mangled key-value mappings:

```ruby
Textbox.includes(:textbox_set)
  .where(textbox_set: { key: 'info', app_scope: 'sgpub' })
  .where(lang: I18n.locale.to_s)
```

Produces:

```SQL
SELECT "textboxes".* FROM "textboxes" 
INNER JOIN "textbox_sets" 
ON "textbox_sets"."id" = "textboxes"."textbox_set_id" 
WHERE "textboxes"."lang" = 'info' 
AND ("textbox_sets"."key" = 'sgpub' 
AND "textbox_sets"."app_scope" = 'en')
```

Changing the order of the where statements, or not chaining them produces correct SQL:

```ruby
Textbox.includes(:textbox_set)
  .where(lang: I18n.locale.to_s)
  .where(textbox_set: { key: 'info', app_scope: 'sgpub' })
```

```ruby
Textbox.includes(:textbox_set)
  .where(textbox_set: { key: 'info', app_scope: 'sgpub' }, lang: I18n.locale.to_s)
```

```SQL
SELECT "textboxes".* FROM "textboxes" 
INNER JOIN "textbox_sets" ON "textbox_sets"."id" = "textboxes"."textbox_set_id" 
WHERE "textboxes"."lang" = 'en' 
AND ("textbox_sets"."key" = 'info' 
AND "textbox_sets"."app_scope" = 'sgpub')
```

Have we stumbled upon a bug, or is this behaviour documented somewhere?

**Comment by sgrif** [comment]
Please provide an executable test case that we can use to reproduce the issue using our [standard bug report template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb)

**Comment by pascalbetz** [comment]
@spiffistan I can not reproduce this. What do your models look like?

**Comment by avand** [comment]
I could not reproduce this issue, although I did find another issue with your original query. I tried with Ruby 2.3.0 and Rails 4.2.5.

First I generated a new Rails app and scaffolded `TextboxSet` and `Textbox`:

```
rails generate scaffold TextboxSet key:string app_scope:string
rails generate scaffold Textbox lang:string textbox_set_id:integer
```

Then I modified the models to account for the relationships between them:

``` ruby
class TextboxSet < ActiveRecord::Base
  has_many :textboxes
end

class Textbox < ActiveRecord::Base
  belongs_to :textbox_set
end
```

Then I attempted the query you originally mentioned was behaving suspiciously:

``` ruby
Textbox.includes(:textbox_set)
  .where(textbox_set: { key: 'info', app_scope: 'sgpub' })
  .where(lang: I18n.locale.to_s)
```

At first, I actually got an exception: `SQLite3::SQLException: no such column: textbox_set.key`. It looks like that query isn't actually correct in the first place. When you use the hash syntax to add a condition to the join table, you'll want to use the pluralized table name (not the model name). So I just changed the `.where(textbox_set...` to `.where(textbox_sets...`. With that change, I get the correct SQL (no incorrect key/value mappings).
